Add greater functionaility to the app

VikramGoyal23 opened the pull request:
### Tyler

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Tyler frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://golfwang.com/).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

##

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Tyler("data/tyler.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

VikramGoyal23 committed:
Rename main class and add greeting

VikramGoyal23 committed:
Add echo functionality

VikramGoyal23 committed:
Add temporary storage functionality

VikramGoyal23 committed:
Add "mark and unmark" functionality and introduce Task class

A "mark and unmark" feature would help users keep track of tasks
better.

Creating a Task class helps to better abstract the project and
assists in the creation of such a feature as mentioned above.

Let's create a Task class with the functionality required to
implement task completion tracking.

VikramGoyal23 committed:
Add Todo, DeadLine and Event

Unique categories of tasks would allow the
users to customise their lists and allow
for greater freedom.

Let's add different categories of Tasks,
namely Todos, Events and Deadlines for the
user to segregate their workflow.

VikramGoyal23 committed:
Update text UI testing

Let's make the UI testing more robust
by testing for more commands.

VikramGoyal23 committed:
Add tests for incorrect inputs

VikramGoyal23 committed:
Add error handling

The user may provide incorrect input which the
program may be unable to process.

Let's add error handling for these edge cases.

VikramGoyal23 committed:
Add delete functionality

Users may need to to delete the tasks in the list
to reduce clutter.

Let's add a delete function.

Luoq1-Xu commented on src/main/java/main/Tyler.java:
```diff
@@ -0,0 +1,99 @@
+
+        Scanner sc = new Scanner(System.in);
+        ArrayList<Task> tasks = new ArrayList<>();
+        int i = 0;
```
This variable seems a little out of place. Perhaps a more informative variable name would be good here.

Luoq1-Xu commented on src/main/java/main/Tyler.java:
```diff
@@ -0,0 +1,99 @@
+                + "\t" + " What can I do for you?\n";
+        String farewell = "\t" + " Bye-bye now!\n";
+        System.out.print(separator + greeting
+                + separator);
```
Perhaps this can be condensed into a single line? It does not seem very long.

Luoq1-Xu commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+     * @param task Task to be added.
+     * @return Next task addition index.
+     * @throws IllegalArgumentException If task description is blank.
+     */
```
Nice javaDocs here, very informative!

Luoq1-Xu commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+package task;
+
+public class Deadline extends Task {
```
Perhaps it would be good to have header comments for each class as well? To give the reader a quick introduction to what the class is about.

Luoq1-Xu reviewed (commented):
Overall looks pretty good! I added some suggestions to make the code slightly more readable. Also, perhaps consider splitting some of the logic in `Tyler.java` into other smaller classes (like a `TaskList `class).

VikramGoyal23 committed:
Add storage functionality

The user may want their tasklist to persist
between sessions.

Let's add a storage function to the bot.

Timothy-Ho0203 commented on src/main/java/main/Tyler.java:
```diff
@@ -0,0 +1,159 @@
+            Files.createFile(tylerPath);
+        }
+
+        List<String> stored = Files.readAllLines(tylerPath);
```
Could this be more specific on what is being stored?

Timothy-Ho0203 reviewed (commented):
Looks good to me, naming of variables are quite compliant with code quality guidelines. There are a few short variables but it's pretty obvious what they mean. Good job!

VikramGoyal23 committed:
Add date and time functionality

The user would benefits from being able to
enter dates in a computer-parsable format, as
it would allow for increased tracking capabilities.

Let's add date and time functionality to the bot,
as well as a command that lists out tasks with
dates matching with the date provided by the user.

VikramGoyal23 committed:
Add more OOP

The current code does not use OOP well.

Let's refactor the codebase to utilize OOP
better for ease of extension in the future.

VikramGoyal23 committed:
Move files to different packages

Let's put all of the files into a package named "tyler".

VikramGoyal23 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

VikramGoyal23 committed:
Add gradle support

Let's add gradle support and modify the build.gradle
file to match the project.

VikramGoyal23 committed:
Add JUnit tests and make small changes

Let's add JUnit tests for ease of future regression testing.

Let's add a catch statement for DateTimeParseException and
change the access of the variables in Task.java to private.

VikramGoyal23 committed:
Modify build.gradle to enable JAR packaging

VikramGoyal23 committed:
Add JavaDocs to multiple classes/methods

Let's add JavaDocs to the *Command classes and methods,
the Task class and its methods, as well as the Ui class
and its methods for better quality documentation.

VikramGoyal23 committed:
Refactor code to comply with coding standard

VikramGoyal23 committed:
Add functionality for find feature

Users may want to search for a task using a keyword in that task.

Let's add a find function to the bot to increase the user's
search capabilities.

VikramGoyal23 committed:
Merge branch 'branch-A-JavaDoc'

VikramGoyal23 committed:
Merge branch 'branch-A-CodingStandard'

VikramGoyal23 committed:
Merge branch 'branch-Level-9'

VikramGoyal23 committed:
Add checkstyle and fix coding style violations.

Let's add support for using checkstyle and fix
some of the coding style violations detected by it.

VikramGoyal23 committed:
Merge branch 'branch-A-CheckStyle'

VikramGoyal23 committed:
Add GUI to bot

The user would prefer a GUI with the
ability to click to interact.

Let's change the bot to work with a GUI.

VikramGoyal23 committed:
Merge branch 'branch-level-10'

VikramGoyal23 committed:
Add assertions to various methods

There are several methods which need to perform
some logical operation. It is important for them
to be correct as otherwise they would cause logic errors.

Adding assertions can help catch logic errors more easily.

As a step towards safer code, let's add assert statements to
critical points of computation in the methods. Doing so will
streamline testing.

VikramGoyal23 committed:
Improve code quality

After rewriting the code for the GUI, the code
needs to be cleaned up to be of better quality.

Let's edit the code to meet the coding standard.

VikramGoyal23 committed:
Merge pull request #2 from VikramGoyal23/branch-A-Assertions

Merge Branch-A-Assertions

VikramGoyal23 committed:
Merge branch 'master' of https://github.com/VikramGoyal23/ip

VikramGoyal23 committed:
Merge pull request #3 from VikramGoyal23/branch-A-CodeQuality

Improve code quality

VikramGoyal23 committed:
Merge branch 'master' of https://github.com/VikramGoyal23/ip

VikramGoyal23 committed:
Add functionality for detecting duplicates

The user might add duplicate tasks, which would cause
unnecessary clutter.

Let's add a functionality to detect when the user is
adding a duplicate task and prevent them from doing so.

VikramGoyal23 committed:
Merge branch 'branch-C-DetectDuplicates'

VikramGoyal23 committed:
Fix detect duplication bug

VikramGoyal23 committed:
Add User Guide

VikramGoyal23 committed:
Fix docs

VikramGoyal23 committed:
Fix uncaught/improperly handled exceptions

Fix the uncaught DateTimeParseException error
in AddCommand for when the user enters the date
in the wrong format.

Fix the try-catch block in DateCommand so that
the app doesn't fail silently when the user uses
the wrong date format.

VikramGoyal23 committed:
Add more abstraction

The code contains a few very long methods.

Let's abstract out the logic to make the code neater.
